### ISSUE TITLE
Tasks are marked as `In Progress` though no builds are scheduled

### ISSUE THREAD

**Issue by keyonghan** [flutter/flutter#122117]
Example from current TOT:

Builder `Linux web_tests_3` is marked as `In Progress` (yellow) for commit https://github.com/flutter/flutter/commit/80be2bfc9f8cc39d197f616b8ce809217908a5f8
<img width="479" alt="Screenshot 2023-03-07 at 10 30 32 AM" src="https://user-images.githubusercontent.com/54558023/223516821-6c886f1e-eeec-4501-b87e-468eef743cee.png">

However, there is no build scheduled per milo page:
<img width="1047" alt="Screenshot 2023-03-07 at 10 30 45 AM" src="https://user-images.githubusercontent.com/54558023/223516937-9457fbf4-8d3e-4ee7-a6fc-83d075201aa2.png">

This happens on random number of tasks intermittently on random commit rows. Should be a bug in the scheduler.

See original findings of this issue: https://github.com/flutter/flutter/issues/120395#issuecomment-1444810718 /cc @ricardoamador

**Comment by keyonghan** [comment]
More context: earlier mitigation was implemented in https://github.com/flutter/cocoon/pull/2471 to vacuum stale tasks, but got due to https://github.com/flutter/flutter/issues/121989.

**Comment by keyonghan** [comment]
Seems related to the https://github.com/flutter/flutter/issues/122116, which caused failures when schedule new builds (this happens after tasks updates). 
At the same time, add more logs to help triage: https://github.com/flutter/cocoon/pull/2518

I would expect the issue mitigated with https://github.com/flutter/cocoon/pull/2516, but needs to monitor.

**Comment by keyonghan** [comment]
Random hanging boxes still repro occassionally:
<img width="533" alt="Screenshot 2023-03-09 at 11 56 47 AM" src="https://user-images.githubusercontent.com/54558023/224140470-e83be1c6-7dee-4575-8874-f39b3a541e57.png">

Looking though backend log, there is indeed no schedule request created. So this is not related to pub/sub message issue.

Digging further.

**Comment by keyonghan** [comment]
Okay, finally found the culprit! 
```
Failed to update tasks when backfilling: Connection closed before full header was received
```

All these hanging tasks happened when triggered via `batch backfiller` API, which failed due to connection close.

**Comment by keyonghan** [comment]
This is due to `eagerError: true` when batch scheduling, which forces connection close when there is an error in any Future. https://github.com/flutter/cocoon/pull/2524 to fix.

**Comment by keyonghan** [comment]
(resetting those boxes to enable backfilling on corresponding columns)

The potential fix https://github.com/flutter/cocoon/pull/2524 is landed, and a new cocoon has been deployed. Keep monitoring to see if any new instances.

**Comment by keyonghan** [comment]
With the new deployment, we got another instance of `Failed to update tasks when backfilling: Connection closed before full header was received`.
<img width="376" alt="Screenshot 2023-03-10 at 1 47 37 PM" src="https://user-images.githubusercontent.com/54558023/224434805-d0bba854-6298-46fc-a00a-dbd4e159a941.png">

**Comment by ricardoamador** [comment]
Is this a connection to Github?

**Comment by keyonghan** [comment]
> Is this a connection to Github?

No. It's calling pubsub APIs to publish numerous messages (several hundred - the number of post submit targets - for the worst case)

I am wondering if too many requests somehow cause connection issue. Two options to limit the number of requests in every API call:
- Limit the number of batch backfilled tasks (see 50 each time)
- Switch one scheduled message on one task to batch scheduling (one message on multiple tasks)
  - batch scheduling is being used to trigger all tasks when new commit merged, which has been working well
  - However, one message per build is preferred https://github.com/flutter/flutter/issues/103443

**Comment by keyonghan** [comment]
Good news, with https://github.com/flutter/cocoon/pull/2529, we are not seeing any instance today. Keep monitoring.

**Comment by keyonghan** [comment]
One side effect is the backfiller keeps backfilling right (dashboard) targets with a "higher" priority (though same), compared to the left side:

As of when posting this comment, the left side:
<img width="723" alt="Screenshot 2023-03-14 at 12 01 06 PM" src="https://user-images.githubusercontent.com/54558023/225110361-4c0fb849-278a-4dd4-a67f-bf11773cdff0.png">

The right side:
<img width="647" alt="Screenshot 2023-03-14 at 12 01 16 PM" src="https://user-images.githubusercontent.com/54558023/225110397-288c777a-49cb-42b0-b1f8-b53efa25a7a2.png">

This is due to similar/same returned task list when querying datastore. We can simply shuffle the list to balance the schedule.

**Comment by keyonghan** [comment]
I am seeing one more instance, but seems a server side error:
```
Failed to update tasks when backfilling: DetailedApiRequestError(status: 503, message: Policy checks are unavailable.)
```

This caused one task to hang. Keep monitoring.

**Comment by keyonghan** [comment]
The instance happened once this morning with connection close...

```
> ERROR 2023-03-16T16:16:16.742Z Failed to update tasks when backfilling: Connection timed out
```

I am wondering if this is related to legacy LUCI buildbucket notifications (http://b/268253534).

**Comment by keyonghan** [comment]
> I am wondering if this is related to legacy LUCI buildbucket notifications (http://b/268253534).

I take this back. This should not be related to LUCI side API, but pub/sub.

**Comment by keyonghan** [comment]
Update: haven't seen a single instance since last week. Moving out of ticket queue for further monitoring.

**Comment by keyonghan** [comment]
Reopen as this issue still happens. Considering tasks are successfully marked as `In Progress`, this should not be related to datastore error. Instead, the builds failed to get scheduled. Though we have [retry logic](https://github.com/flutter/cocoon/pull/2589) implemented, we do not seem to catch all errors.

<img width="1375" alt="Screenshot 2023-11-08 at 2 23 35 PM" src="https://github.com/flutter/flutter/assets/54558023/4f6c8cd4-ec8f-4a6c-8605-555b88ab97cf">

**Comment by keyonghan** [comment]
https://github.com/flutter/cocoon/pull/3242 to catch all error exceptions to retry. Waiting to see if this fixes/mitigate the problem.

**Comment by ricardoamador** [comment]
Keyong is out until December, so I am monitoring the fix. Reducing priority as this does not seem to be happening with submitted change.

**Comment by ricardoamador** [comment]
Unfortunately this is still happening.
![image](https://github.com/flutter/flutter/assets/32242716/0d78d3a9-c34e-4530-8a46-9a7525c8ee1d)

**Comment by ricardoamador** [comment]
Okay this is actually working! We need to look at the default values for the dashboard to fix the presentation of tasks so they do not show with that high queue time.

**Comment by ricardoamador** [comment]
Closing as this is working.
